### PR TITLE
Bell/hetero execution devices

### PR DESCRIPTION
**Pull request by songbell**
### Details:
 - support execution_device in hetero

### Tickets:
 - 92661

### CODE REVIEW

File: src/plugins/hetero/executable_network.cpp [hunk 1]
```diff
@@ -515,7 +515,9 @@ HeteroExecutableNetwork::HeteroExecutableNetwork(const InferenceEngine::CNNNetwo
         network._network = _heteroPlugin->GetCore()->LoadNetwork(network._clonedNetwork,
                                                                  network._device,
                                                                  metaDevices[network._device]);
+        _exeDevices += network._device + ",";
     }
+    _exeDevices.pop_back();
```
Comment (songbell) [inline src/plugins/hetero/executable_network.cpp:520]: @ilya-lavrenov for hetero, the exedevice, shall we list all the targets for sub networks, like GPU,CPU,GPU,CPU? or we only keep one occurrence of the target hw?

Comment (nkogteva) [reply src/plugins/hetero/executable_network.cpp:520]: @songbell @ilya-lavrenov I think we should avoid duplicates and keep execution device one time. Also maybe there is a sense to keep order according to priority.

Comment (ilya-lavrenov) [reply src/plugins/hetero/executable_network.cpp:520]: agree with @nkogteva 
also, why do we need to keep it as a field of class? I suppose we can create it on demand once user requested it.
Secondly, why `ov::execution_devices` is a string? does it make sense to have it to be `std::vector<std::string>`?

Comment (songbell) [reply src/plugins/hetero/executable_network.cpp]: changed the implementation,  please review again @ilya-lavrenov  @nkogteva

File: samples/cpp/benchmark_app/main.cpp [hunk 2]
```diff
@@ -1253,8 +1253,11 @@ int main(int argc, char* argv[]) {
             statistics->dump();
 
         // Performance metrics report
-        if (device_name.find("AUTO") != std::string::npos)
-            slog::info << "ExecutionDevice:   " << compiledModel.get_property(ov::execution_devices) << slog::endl;
+        try {
+            auto exeDevice = compiledModel.get_property(ov::execution_devices);
+            slog::info << "ExecutionDevice:   " << exeDevice << slog::endl;
+        } catch (...) {
```
Comment (nkogteva) [inline samples/cpp/benchmark_app/main.cpp:1259]: Maybe catch specific exception just to skip non implemented only?

File: src/plugins/hetero/executable_network.cpp [hunk 3]
```diff
@@ -957,6 +957,20 @@ InferenceEngine::Parameter HeteroExecutableNetwork::GetMetric(const std::string&
...
+    } else if (name == ov::execution_devices) {
+        std::vector<std::string> exeDevices;
+        for (auto&& subnetwork : _networks)
+            exeDevices.push_back(subnetwork._device);
+        std::set<std::string> s;
+        for (auto iter = exeDevices.begin(); iter != exeDevices.end();) {
+            if (s.count(*iter) == 0) {
+                s.insert(*iter);
+            } else {
+                exeDevices.erase(iter--);
```
[the remark this answers is not in the document]
Comment (songbell) [reply src/plugins/hetero/executable_network.cpp]: fixed.

File: samples/cpp/benchmark_app/main.cpp [hunk 4]
```diff
@@ -1253,8 +1253,11 @@ int main(int argc, char* argv[]) {
             statistics->dump();
 
         // Performance metrics report
-        if (device_name.find("AUTO") != std::string::npos)
-            slog::info << "ExecutionDevice:   " << compiledModel.get_property(ov::execution_devices) << slog::endl;
+        try {
+            auto exeDevice = compiledModel.get_property(ov::execution_devices);
+            slog::info << "ExecutionDevice:   " << exeDevice << slog::endl;
```
[the remark this answers is not in the document]
Comment (songbell) [reply samples/cpp/benchmark_app/main.cpp]: fixed

File: samples/cpp/benchmark_app/main.cpp (hunk 2 above)
Comment (songbell) [reply samples/cpp/benchmark_app/main.cpp]: fixed

File: src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp [hunk 5]
```diff
@@ -345,6 +346,25 @@ TEST_P(OVClassHeteroExecutableNetworkGetMetricTest_TARGET_FALLBACK, GetMetricNoT
...
+    }
+    for (auto& node : actualNetwork->get_ops()) {
+        auto affinity = layermap[node->get_friendly_name()];
+        node->get_rt_info()["affinity"] = affinity;
+    }
+    auto compiled_model = ie.compile_model(actualNetwork, heteroDeviceName);
+
+    std::vector<std::string> exeTargets;
+    OV_ASSERT_NO_THROW(exeTargets = compiled_model.get_property(ov::execution_devices));
+    std::vector<std::string> expectedTargets = {target_device, CommonTestUtils::DEVICE_CPU};
```
[the remark this answers is not in the document]
Comment (nkogteva) [reply src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp:364]: > yes, this case set affinity with some nodes on CPU

When we run on CPU we should have `CPU` at once in expected targets, is not it?

[the remark this answers is not in the document]
Comment (songbell) [reply src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp]: you mean -d HETERO:CPU？ in this case ,the expected targets is CPU only.
in the case, it's -d HETERO:GPU,CPU with subnetworks running on cpu, so the expected targets is GPU,CPU

[the remark this answers is not in the document]
Comment (nkogteva) [reply src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp]: > you mean -d HETERO:CPU？ in this case ,the expected targets is CPU only. in the case, it's -d HETERO:GPU,CPU with subnetworks running on cpu, so the expected targets is GPU,CPU

Yes, I'm talking about case when we have just HETERO:CPU

File: tools/benchmark_tool/openvino/tools/benchmark/main.py [hunk 6]
```diff
@@ -575,8 +575,11 @@ def set_throughput_streams():
                                       ])
             statistics.dump()
 
-        if devices.count("AUTO"):
-            print(f'ExecutionDevice: {compiled_model.get_property("EXECUTION_DEVICES")}')
+        try:
+            exeDevice = compiled_model.get_property("EXECUTION_DEVICES")
+            print(f'ExecutionDevice: {exeDevice}')
```
[the remark this answers is not in the document]
Comment (songbell) [reply tools/benchmark_tool/openvino/tools/benchmark/main.py]: fixed

File: src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_executable_network/core_integration.cpp [hunk 7]
```diff
@@ -75,7 +75,9 @@ INSTANTIATE_TEST_SUITE_P(
 //INSTANTIATE_TEST_SUITE_P(
 //        smoke_OVClassHeteroExecutableNetworkGetMetricTest, OVClassHeteroExecutableNetworkGetMetricTest_TARGET_FALLBACK,
 //        ::testing::Values("CPU"));
-
+INSTANTIATE_TEST_SUITE_P(
+        smoke_OVClassHeteroExecutableNetworkGetMetricTest, OVClassHeteroExecutableNetworkGetMetricTest_EXEC_DEVICES,
+        ::testing::Values("CPU"));
```
Comment (songbell) [inline src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_executable_network/core_integration.cpp:80]: HETERO:CPU case is covered here @nkogteva

**Review by nkogteva — APPROVED** [review]
LGTM

File: src/inference/include/openvino/runtime/properties.hpp [hunk 8]
```diff
@@ -889,5 +889,5 @@ static constexpr Property<Affinity> affinity{"AFFINITY"};
  * @brief The devices that the inference task been executed.
  * @ingroup ov_runtime_cpp_prop_api
  */
-static constexpr Property<std::string> execution_devices{"EXECUTION_DEVICES"};
+static constexpr Property<std::vector<std::string>, PropertyMutability::RO> execution_devices{"EXECUTION_DEVICES"};
```
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/inference/include/openvino/runtime/properties.hpp:892]: yes, we have not released it yet. It's in master only

**Review by ilya-lavrenov — APPROVED** [review]

**Review by wangleis — APPROVED** [review]

File: src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp [hunk 9]
```diff
@@ -345,6 +346,29 @@ TEST_P(OVClassHeteroExecutableNetworkGetMetricTest_TARGET_FALLBACK, GetMetricNoT
     ASSERT_EQ(expectedTargets, targets);
 }
 
+TEST_P(OVClassHeteroExecutableNetworkGetMetricTest_EXEC_DEVICES, GetMetricNoThrow) {
+    ov::Core ie = createCoreWithTemplate();
+    std::vector<std::string> expectedTargets = {target_device};
+#ifdef ENABLE_INTEL_CPU
```
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp:352]: we must not have plugin dependent test in common part. Could you please rework the tests to be sustainable and not plugin dependent?

Comment (songbell) [reply src/tests/functional/plugin/shared/include/behavior/ov_executable_network/get_metric.hpp:352]: @ilya-lavrenov will work on this. @wgzintel please noted on this requirement as well

**Pull request merged**